client-app/src/examples/todo/TodoPanelModel.js

```diff
@@ -0,0 +1,268 @@
+                        Icon.circle({prefix: 'fal', className: 'xh-text-color-muted'}) :
+                        Icon.checkCircle({prefix: 'fal', className: 'xh-intent-success'}),
+                    tooltip: text
+                } :
+                {hidden: true};
+        },
+        actionFn: ({selectedRecords}) => {
+            const tasks = selectedRecords.map(it => it.data);
+
+            this.toggleCompleteAsync(tasks);
+        }
+    });
+
+    get selectedTasks() {
+        return this.gridModel.selection.map(it => it.data);
+    }
+
+    constructor() {
+        super();
+        makeObservable(this);
+
+        this.gridModel = this.createGridModel();
+
+        this.addReaction({
+            track: () => this.showCompleted,
+            run: (showCompleted) => {
+                const filter = showCompleted ? null : {field: 'complete', op: '=', value: false};
+                this.gridModel.setFilter(filter);
+            },
+            fireImmediately: true
+        });
+
+        this.addReaction({
+            track: () => this.showGroups,
+            run: (showGroups) => {
+                this.gridModel.setGroupBy(showGroups ? 'dueDateGroup' : null);
+                this.gridModel.hideColumn('dueDateGroup');
+            },
+            fireImmediately: true
+        });
+    }
+
+    async addTaskAsync(task) {
+        await XH.todoService.addTaskAsync(task);
+        await this.refreshAsync();
+        this.info(`Task added: '${task.description}'`);
+    }
+
+    async editTaskAsync(task) {
+        await XH.todoService.editTaskAsync(task);
+        await this.refreshAsync();
+        this.info(`Task edited: '${task.description}'`);
+    }
+
+    async toggleCompleteAsync(tasks) {
+        if (isEmpty(tasks)) return;
+
+        const firstTask = tasks[0],
+            isCompleting = !firstTask.complete;
+
+        await XH.todoService.toggleCompleteAsync(tasks);
+        await this.refreshAsync();
+
+        if (isCompleting) {
+            const count = tasks.length;
+            if (count > 1) {
+                this.info(`Congrats! You completed ${count} tasks!`);
+            } else {
+                this.info(`Congrats! You completed '${firstTask.description}!'`);
+            }
+        }
+    }
+
+    async removeSelectedTasksAsync() {
+        const {selectedTasks} = this;
+
+        if (isEmpty(selectedTasks)) return;
+
+        const count = selectedTasks.length,
+            {description} = selectedTasks[0],
+            message = count === 1 ? `'${description}?'` : `${count} tasks?`,
+            confirmed = await XH.confirm({
+                title: 'Confirm',
+                message: `Are you sure you want to permanently remove ${message}`
+            });
+
+        if (!confirmed) return;
+
+        const label = count === 1 ? `Task removed: '${description}'` : `${count} tasks removed`;
+
+        for (const task of selectedTasks) {
+            await XH.todoService.removeTasksAsync(task);
```

Comment:
pluralization in this method name is wrong, as it currently only does one - but given how we are actually using it here, would make sense to actually go ahead and implement the bulk delete at the service 

In a real world system that would be a lot safer -- want to modify state/backend in one shot whenever it is easier to do so -- we can discuss